Fixed segfault when an NPC completes a pickup after a player has alre…

Levitator0 opened the pull request:
abusive off-topic commentary removed

Commit:
Fixed segfault when an NPC completes a pickup after a player has already taken it

- Fixes https://github.com/CleverRaven/Cataclysm-DDA/issues/72340
- Replaces naked pointer npc::wanted_item with: item_location npc::wanted_item
- The safe reference will return null if the object gets deleted, which is what happens
  if someone else collects it before the NPC completes the pickup

TODO: Sensibly, we should track the item coordinates using item_location
since it's suppose to package a safe reference alongside location data.
However, npc::wanted_item_pos is subject to shuffling via the game::update_map()
and shift() mechanism, so rather than try to comprehend that and how it
does or doesn't interact with item_location,  I'm instead leaving
it alone since the coordinates appear to be calculated right at present.

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Levitator0 commented on src/npcmove.cpp:
```diff
@@ -3523,23 +3525,26 @@
         }
+        else
+            return false;
     };
```
Wut? Not enough context to be sure it's not mismatching braces.

mqrause replied:
This is definitely missing the closing brace of the `else`. And this is also causing all the following style suggestions of code you didn't touch.

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Levitator0 commented on src/npcmove.cpp:
```diff
@@ -3523,23 +3525,25 @@ void npc::find_item()
+            return true;
+        } else {
+            return false;
     };
```
wut?

Levitator0 commented on src/npcmove.cpp:
```diff
@@ -3577,16 +3581,17 @@
+                if( consider_item( it, p ) )
+                    wanted_item = item_location{ map_cursor{p}, &it };
             }
         }
```
?

Levitator0 commented on src/npcmove.cpp:
```diff
@@ -3577,16 +3581,17 @@
             }
         };
         bool can_see = false;
         if( here.sees_some_items( p, *this ) && sees( p ) ) {
             can_see = true;
-            for( const item &it : m_stack ) {
-                consider_item( it, p );
+            for( item &it : m_stack ) {
+                if( consider_item( it, p ) )
+                    wanted_item = item_location{ map_cursor{p}, &it };
```
Not going to commit any of these I don't understand.

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

mqrause commented:
Those suggested astyle changes strongly indicate you got some missing or extra braces.

Also please remove unnecessary code instead of just commenting it out. If you think a justification of the removal is necessary, put it in the PR text.

Levitator0 commented:
> Those suggested astyle changes strongly indicate you got some missing or extra braces.
> 
> Also please remove unnecessary code instead of just commenting it out. If you think a justification of the removal is necessary, put it in the PR text.

There was an ubraced single-line else statement, so maybe that's the issue. I don't think I know how to read some of those prompts, so I didn't commit them. If the devs want to assume style changes are always valid, they're welcome to apply them.

mqrause commented:
Please check the [style guide](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.md) for how we make sure everyone uses the same code style. Commits become very messy very fast is everyone uses their own favorite style.
As a rule of thumb, astyle is always right. And if code you didn't touch fails astyle checks, it's almost certainly an error in the code you did touch.

And as another rule of thumb, please don't mark suggestions as resolved when you're unsure about why they're there and what they try to achieve.

Commit:
Update src/npcmove.cpp

single-line else needing both braces

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Levitator0 commented:
> Please check the [style guide](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.md) for how we make sure everyone uses the same code style. Commits become very messy very fast if everyone uses their own favorite style. As a rule of thumb, astyle is always right. And if code you didn't touch fails astyle checks, it's almost certainly an error in the code you did touch.
> 
> And as another rule of thumb, please don't mark suggestions as resolved when you're unsure about why they're there and what they try to achieve.
> 
> Edit: I left a comment on the third review, in case you didn't see that.

I'm beyond confused as to how this issue resolution system interacts with the lint bot. I have totally lost the thread. It seems like the number of issues to resolve has only climbed. It doesn't upfront distinguish between an issue that was closed accepted/committed or closed as disregarded. And now I have a bunch of style issues, where when I attempt to commit them, GitHub complains that edits on deleted lines are not supported. Well, then why is it proposing the commit? I have no idea how to use this arrangement. I'm going to try to lint the code locally and then either update the PR, or reissue it, because this is driving me crazy.

Commit:
Locally linted npcmove.cpp

This is my first submission, and I ran into issues with the GitHub linter,
so hopefully this is way simpler.

Levitator0 commented:
Ok, I linted locally, and I have to conclude that astyle flips out and becomes totally confused if you leave the braces off of a single-line else statement. In any case, the code passes the automated check on GitHub now, so I'm going to mark all of the linting conversations as resolved because linting passes both locally and online. Sorry for the mess, as this setup confused the daylights out of me.

NetSysFire commented:
1. You can batch apply suggestions.
2. You still need to fix your PR summary.

Also dear mergers, please squash this one.

Levitator0 commented:
> 1. You can batch apply suggestions.
> 
>     2. You still need to fix your PR summary.
> 
> 
> Also dear mergers, please squash this one.

What's wrong with my summary?

NetSysFire commented:
The summary section must contain exactly one line, in the format of Category "Description here". See test failure of the PR validator test. The category to use here is Bugfixes. Move your closing keyword into the Purpose of change section.

If you stayed just 10 minutes longer on IRC I would have helped you there.

Commit:
Fixed a complaint about if() braces from clang-tidy

Commit:
Update src/npcmove.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Fixed another if() brace complaint for clang-tidy

Commit:
astyle demanded a space if() {
                            ^

I-am-Erk commented on src/npc.h:
```diff
@@ -1358,7 +1358,12 @@ class npc : public Character
+
+        //Safe reference to an item at a specific location which can be checked for subsequent invalidation
+        //in case the object is destroyed or more than one creature goes for it.
+        //const item *wanted_item = nullptr;
```
You're overdoing it on the comments here (which is better than underdoing it but still should be addressed). The very fact that you're using an item_location basically tells us what you're saying in the comments. You could say just `// make sure to use item_location in case something happens to the item` if you want an abundance of clarity, but I don't think a comment is necessary at alll

Levitator0 replied:
I haven't seen it documented where it's explained what conditions invalidate an item_location. One would hope that location movement would do it, but if movement is defined as deletion and re-creation then it's the same as for deletion. There's a actually a lot to be said about item_location, but maybe I haven't looked in the right place.

I-am-Erk commented on src/npcmove.cpp:
```diff
@@ -3523,23 +3525,26 @@ void npc::find_item()
-            wanted_item = &it;
-            wanted = wanted_item;
+            //wanted_item = &it;
+            //wanted = wanted_item;
```
This is dead code now, don't comment it out. You'll confuse us simple folk when we come to edit this code in seven years time.

I-am-Erk reviewed (approved):
To my limited skill the code looks good, just some comments on comments.

Commit:
Comment cleanup per I-am-Erk

mqrause commented on src/npcmove.cpp:
```diff
@@ -3514,7 +3515,8 @@ void npc::find_item()
     //range = std::max( 1, std::min( 12, range ) );
 
-    const item *wanted = wanted_item;
+    //Redundant with respect to wanted_item
+    //item_location wanted = wanted_item;
```
Erk missed that one, that should also go.

Commit:
Removed some more comments

Zireael07 commented:
Why close?

Levitator0 commented:
> abusive off-topic commentary removed

No, you know what's abusive? It's having a bunch of second-rate flunkies treat programmers and contributors with disrespect. I'm not your bitch. You have to shell out good money for someone to whore themselves to you and smile when you tell them to shut their mouth all day. I'll make sure Kevin knows about this so he never wonders why there's a shortage of C++ skils.

Levitator0 commented:
> Why close?

Because kevin lets a bunch of dicks run his shop.

ZhilkinSerg commented:
> Why close?

I do not think they want to contribute.
